File: Python/Lesson_084_kth_smallest_element_in_a_bst.py
# =============================================================
# MIT License | @analyticswithharry2026
# GitHub  : https://github.com/analyticswithharry
# YouTube : Analytics with Harry
# =============================================================
# Lesson     : 084 -- Kth Smallest Element in a BST
# Category   : Trees
# Difficulty : Medium
# Study Plan : Day 42
# =============================================================
#
# QUESTION:
#   Given a BST, return the kth smallest value (1-indexed). Use inorder traversal.
# =============================================================

class TreeNode:
    def __init__(self, v=0, l=None, r=None): self.val=v; self.left=l; self.right=r
class Solution:
    def kthSmallest(self, root, k):
        st, cur = [], root
        while st or cur:
            while cur: st.append(cur); cur = cur.left
            cur = st.pop(); k -= 1
            if k == 0: return cur.val
            cur = cur.right

if __name__ == "__main__":
    r = TreeNode(3, TreeNode(1, None, TreeNode(2)), TreeNode(4))
    print(Solution().kthSmallest(r, 1))
    print(Solution().kthSmallest(r, 3))

File: CPP/Lesson_084_kth_smallest_element_in_a_bst.cpp
// =============================================================
// MIT License | @analyticswithharry2026
// GitHub  : https://github.com/analyticswithharry
// YouTube : Analytics with Harry
// =============================================================
// Lesson     : 084 -- Kth Smallest Element in a BST
// Category   : Trees
// Difficulty : Medium
// Study Plan : Day 42
// =============================================================
//
// QUESTION:
//   Given a BST, return the kth smallest value (1-indexed). Use inorder traversal.
// =============================================================

#include <vector>
#include <string>
#include <iostream>
#include <stack>
#include <queue>
#include <unordered_map>
#include <unordered_set>
#include <map>
#include <set>
#include <algorithm>
#include <climits>
#include <numeric>
#include <functional>
#include <cmath>
using namespace std;
struct TreeNode { int val; TreeNode *left, *right; TreeNode(int v):val(v),left(nullptr),right(nullptr){} };
class Solution { public:
    int kthSmallest(TreeNode* root, int k){
        stack<TreeNode*> st; TreeNode* cur = root;
        while (!st.empty() || cur){
            while (cur){ st.push(cur); cur = cur->left; }
            cur = st.top(); st.pop(); if (--k == 0) return cur->val;
            cur = cur->right;
        }
        return -1;
    }
};
int main(){ auto r = new TreeNode(3); r->left = new TreeNode(1); r->right = new TreeNode(4);
    r->left->right = new TreeNode(2);
    cout<<Solution().kthSmallest(r,1)<<" "<<Solution().kthSmallest(r,3)<<endl; }
